src/cdk/stepper/stepper.ts

```diff
@@ -84,6 +97,10 @@ export class CdkStepper {
   @Input()
   get selectedIndex() { return this._selectedIndex; }
   set selectedIndex(index: number) {
+    this._steps.toArray()[this._selectedIndex].interacted = true;
+    for (let i = 0; i < index; i++) {
+      if (!this._steps.toArray()[i].stepControl.valid) { return; }
```

Comment:
Nit: Here you're calling `toArray()` every index of the loop. Perhaps save it at the top instead? 

```ts
const steps = this._steps.toArray();
steps[this._selectedIndex].interacted = true;
for (let i =0; i < index; i++) {
  if (!steps[i].stepControl.valid) { return; }
}
```

I'd also consider maybe breaking this loop out into its own mini-method, like `_anyControlsInvalid`.